crates/aranya-runtime/src/client.rs

```
@@ -142,18 +142,26 @@ where
         trx: &mut Transaction<SP, E>,
         sink: &mut impl Sink<E::Effect>,
         commands: &[impl Command],
-        request_heads: &mut PeerCache,
     ) -> Result<usize, ClientError> {
-        let count = trx.add_commands(
-            commands,
-            &mut self.provider,
-            &mut self.engine,
-            sink,
-            request_heads,
-        )?;
+        let count = trx.add_commands(commands, &mut self.provider, &mut self.engine, sink)?;
         Ok(count)
     }
 
+    pub fn update_heads(
+        &mut self,
+        storage_id: GraphId,
+        commands: &[impl Command],
+        request_heads: &mut PeerCache,
+    ) -> Result<(), ClientError> {
+        let storage = self.provider.get_storage(storage_id)?;
+        for command in commands {
+            if let Some(loc) = storage.get_location(command.address()?)? {
```

Is there a case where this being `None` isn't an error?

---

IOW should we treat a missing command as an error instead of silently ignoring it.

---

One of the times we update heads is when we get a sample from the peers graph. These commands may not be in our graph. I wasn't using `update_heads` in that case. I updated these code to also call `update_heads` then. Let me know if you think the errors are important and I should separate them.

---

Oh I see, I think this makes sense then.